watches.go

```diff
@@ -52,39 +67,72 @@ func (wm *watchManager) handleWatchRequest(w http.ResponseWriter, r *http.Reques
 		http.Error(w, fmt.Sprintf("invalid 'fromIndex': %v", err), http.StatusBadRequest)
 		return
 	}
-
-	go wm.manageWatch(conn, topic, genID, idx)
+	log.Printf("Connection accepted from %v", r.Host)
+	err = wm.manageWatch(w, topic, genID, idx)
+	if err != nil {
+		log.Printf("Error: handling messages %v\n", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
 }
 
-func (wm *watchManager) manageWatch(conn *websocket.Conn, topic, genID string, idx uint64) {
-	log.Printf("Connection accepted from %v", conn.RemoteAddr())
-	defer closeConn(conn)
-	for {
-		msgsResponse, err := wm.store.get(topic, genID, idx)
-		if err != nil {
-			handleError(err, conn)
-			return
+func (wm *watchManager) manageWatch(w http.ResponseWriter, topic string, genID string, idx uint64) error {
+	var err error
+	var closed bool
+	cn, canNotifyClose := w.(http.CloseNotifier)
+	flusher, canFlush := w.(http.Flusher)
+	if canNotifyClose && canFlush {
+		for aw := newActiveWatch(wm, topic, genID, idx, httputil.NewChunkedWriter(w), flusher); ; {
+			select {
+			case <-cn.CloseNotify():
+				aw.close()
+			default:
+				err = aw.handleNewMessages()
+			}
+			if closed || err != nil {
+				break
+			}
+			time.Sleep(wm.pushInterval)
 		}
+	} else {
+		err = errors.New("Error: cannot stream")
+	}
+	return err
+}
+
+func (aw *activeWatch) handleNewMessages() error {
+	var err error
+	var msgsResponse *MessagesResponse
+	if msgsResponse, err = aw.newMessages(); err == nil {
 		if msgsLength := len(msgsResponse.Messages); msgsLength > 0 {
-			if err := conn.WriteJSON(msgsResponse); err != nil {
-				handleError(err, conn)
-				return
+			if err := aw.writeChunk(msgsResponse); err == nil {
```

Comment:
see my nit comment about error checking in `*activeWatch.writeChunk`, the same applies here